undofx/src/main/java/org/fxmisc/undo/impl/UndoManagerImpl.java

```diff
@@ -209,7 +229,7 @@ private void addChange(C change) {
     @SuppressWarnings("unchecked")
     private C[] merge(C c1, C c2) {
         Optional<C> merged = merge.apply(c1, c2);
-        if(merged.isPresent()) {
+        if(merged.isPresent() && isNotIdentity.test(merged.get())) {
             return (C[]) new Object[] { merged.get() };
         } else {
             return (C[]) new Object[] { c1, c2 };
```

Comment:
One thought I had about the `Either<C, C>` approach. When I implement the non-linear PR, wouldn't it make sense to have a `preventBubble` method? If so, how should that meta-data be stored? Then a change could be

mergeable | bubble-able
--------------|--------------
x | x
x | _
_ | x
_ | _

If we follow the `Either<Barrier, C>` approach, `Barrier` could store all three: the change itself, whether it can be merged, and whether it can be bubbled.